File: packages/python/src/prefactor_evals_no_llm/evals/core/loop_detection.py
"""core.loop_detection

Spec: spec/evals/core/loop_detection.md
"""

from __future__ import annotations

from collections import defaultdict

from ...helpers import digest, span_ids, strip_keys, tool_signature
from ...registry import register
from ...result import failed, passed, skipped
from ...config import cfg_int, cfg_list

EVAL_ID = "core.loop_detection"


@register(EVAL_ID, pack="core")
def run(instance, config, context):
    max_occurrences = cfg_int(config, "max_occurrences", 3)
    ignore_tools = cfg_list(config, "ignore_tools")
    ignore_arg_keys = config.get("ignore_arg_keys") or ()

    tool_calls = [s for s in instance.spans_of_type("tool_call")
                  if s.name not in ignore_tools]
    if not tool_calls:
        return skipped(
            EVAL_ID, instance.id,
            "No tool_call spans on this instance, nothing to check.",
            remedy=("If this agent does call tools, its span schema names are not "
                    "recognised. Map them with span_type_map in the pack file."),
        )

    groups = defaultdict(list)
    for span in tool_calls:
        groups[tool_signature(span, ignore_arg_keys)].append(span)

    offenders = [(sig, spans) for sig, spans in groups.items()
                 if len(spans) > max_occurrences]

    if not offenders:
        worst = max((len(v) for v in groups.values()), default=0)
        return passed(
            EVAL_ID, instance.id,
            "No tool signature repeated more than %d times." % max_occurrences,
            values={"max_occurrences": max_occurrences, "worst_count": worst},
        )

    # Deterministic ordering: worst first, then by tool name for stable ties.
    offenders.sort(key=lambda item: (-len(item[1]), item[1][0].name))
    worst_spans = offenders[0][1]

    evidence_spans = [s for _, spans in offenders for s in spans]
    return failed(
        EVAL_ID, instance.id,
        'Tool "%s" called %d times with identical arguments, limit is %d.'
        % (worst_spans[0].name, len(worst_spans), max_occurrences),
        span_ids=span_ids(evidence_spans),
        values={
            "max_occurrences": max_occurrences,
            "offenders": [
                {
                    "tool": spans[0].name,
                    "count": len(spans),
                    "limit": max_occurrences,
                    # A hash, not the raw arguments: arguments can be large and
                    # can contain sensitive values. They are one span lookup
                    # away in Prefactor.
                    "arg_digest": digest(strip_keys(spans[0].input, ignore_arg_keys)),
                }
                for _, spans in offenders
            ],
        },
    )

File: packages/typescript/src/checks/loopDetection.ts
/**
 * core.loop_detection
 *
 * Spec: spec/evals/core/loop_detection.md
 */

import type { Config } from '../config.js';
import { cfgInt, cfgList } from '../config.js';
import { comparePython } from '../fmt.js';
import { digest, spanIds, stripKeys, toolSignature } from '../helpers.js';
import type { EvalResult } from '../result.js';
import { failed, passed, skipped } from '../result.js';
import type { EvalInstance, EvalSpan } from '../schema.js';
import { spansOfType } from '../schema.js';
import type { EvalContext } from '../registry.js';

export const EVAL_ID = 'core.loop_detection';

export function run(
  instance: EvalInstance,
  config: Config,
  _context: EvalContext,
): EvalResult | null {
  const maxOccurrences = cfgInt(config, 'max_occurrences', 3)!;
  const ignoreTools = cfgList(config, 'ignore_tools');
  const rawIgnoreArgKeys = config['ignore_arg_keys'];
  const ignoreArgKeys: string[] = Array.isArray(rawIgnoreArgKeys)
    ? (rawIgnoreArgKeys as string[]) : [];

  const toolCalls = spansOfType(instance, 'tool_call')
    .filter((s) => !ignoreTools.includes(s.name));
  if (toolCalls.length === 0) {
    return skipped(
      EVAL_ID, instance.id,
      'No tool_call spans on this instance, nothing to check.',
      'If this agent does call tools, its span schema names are not '
      + 'recognised. Map them with span_type_map in the pack file.',
    );
  }

  // A Map, not an object: object keys that look like integers would reorder,
  // and the group order has to match Python's insertion ordered dict.
  const groups = new Map<string, EvalSpan[]>();
  for (const span of toolCalls) {
    const signature = toolSignature(span, ignoreArgKeys);
    const group = groups.get(signature);
    if (group) group.push(span);
    else groups.set(signature, [span]);
  }

  const offenders = [...groups.entries()].filter(([, spans]) => spans.length > maxOccurrences);

  if (offenders.length === 0) {
    let worst = 0;
    for (const spans of groups.values()) worst = Math.max(worst, spans.length);
    return passed(
      EVAL_ID, instance.id,
      `No tool signature repeated more than ${maxOccurrences} times.`,
      undefined,
      { max_occurrences: maxOccurrences, worst_count: worst },
    );
  }

  // Deterministic ordering: worst first, then by tool name for stable ties.
  offenders.sort((a, b) => (b[1].length - a[1].length)
    || comparePython(a[1][0]!.name, b[1][0]!.name));
  const worstSpans = offenders[0]![1];

  const evidenceSpans = offenders.flatMap(([, spans]) => spans);
  return failed(
    EVAL_ID, instance.id,
    `Tool "${worstSpans[0]!.name}" called ${worstSpans.length} times with identical `
    + `arguments, limit is ${maxOccurrences}.`,
    spanIds(evidenceSpans),
    {
      max_occurrences: maxOccurrences,
      offenders: offenders.map(([, spans]) => ({
        tool: spans[0]!.name,
        count: spans.length,
        limit: maxOccurrences,
        // A hash, not the raw arguments: arguments can be large and can
        // contain sensitive values. They are one span lookup away in
        // Prefactor.
        arg_digest: digest(stripKeys(spans[0]!.input, ignoreArgKeys)),
      })),
    },
  );
}
